drivers/video/ov5640.c

```diff
@@ -0,0 +1,670 @@
+	uint16_t chip_id;
+	int ret;
+
+	/* Power up sequence */
+#if DT_INST_NODE_HAS_PROP(0, powerdown_gpios)
+	ret = gpio_pin_configure_dt(&cfg->powerdown_gpio, GPIO_OUTPUT_ACTIVE);
+	if (ret) {
+		return ret;
+	}
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	ret = gpio_pin_configure_dt(&cfg->reset_gpio, GPIO_OUTPUT_ACTIVE);
+	if (ret) {
+		return ret;
+	}
+#endif
+
+	k_sleep(K_MSEC(5));
+
+#if DT_INST_NODE_HAS_PROP(0, powerdown_gpios)
+	gpio_pin_set_dt(&cfg->powerdown_gpio, 0);
+#endif
+
+	k_sleep(K_MSEC(1));
+
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	gpio_pin_set_dt(&cfg->reset_gpio, 0);
+#endif
+
+	k_sleep(K_MSEC(20));
+
+	/* Software reset */
+	ret = ov5640_write_reg(&cfg->i2c, SYS_CTRL0_REG, SYS_CTRL0_SW_RST);
+	if (ret) {
+		LOG_ERR("Unable to perform software reset");
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(5));
+
+	/* Initialize register values */
+	ret = ov5640_write_multi_regs(&cfg->i2c, ov5640InitParams, ARRAY_SIZE(ov5640InitParams));
+	if (ret) {
+		LOG_ERR("Unable to initialize the sensor");
+		return -EIO;
+	}
+
+	/* Set virtual channel */
+	ret = ov5640_modify_reg(&cfg->i2c, 0x4814, 3U << 6, (uint8_t)(DEFAULT_MIPI_CHANNEL) << 6);
+	if (ret) {
+		LOG_ERR("Unable to set virtual channel");
+		return -EIO;
+	}
+
+	/* Check sensor chip id */
+	ret = ov5640_read_reg(&cfg->i2c, CHIP_ID_REG, &chip_id, sizeof(chip_id));
+	if (ret) {
+		LOG_ERR("Unable to read sensor chip ID, ret = %d", ret);
+		return -ENODEV;
+	}
+
+	if (chip_id != CHIP_ID_VAL) {
+		LOG_ERR("Wrong chip ID: %04x (expected %04x)", chip_id, CHIP_ID_VAL);
+		return -ENODEV;
+	}
+
+	/* Set default format to 720p RGB565 */
+	fmt.pixelformat = VIDEO_PIX_FMT_RGB565;
+	fmt.width = 1280;
+	fmt.height = 720;
+	fmt.pitch = fmt.width * 2;
+	ret = ov5640_set_fmt(dev, VIDEO_EP_OUT, &fmt);
+	if (ret) {
+		LOG_ERR("Unable to configure default format");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static const struct ov5640_config ov5640_cfg_0 = {
+	.i2c = I2C_DT_SPEC_INST_GET(0),
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	.reset_gpio = GPIO_DT_SPEC_INST_GET(0, reset_gpios),
+#endif
+#if DT_INST_NODE_HAS_PROP(0, powerdown_gpios)
+	.powerdown_gpio = GPIO_DT_SPEC_INST_GET(0, powerdown_gpios),
+#endif
+};
+
+static struct ov5640_data ov5640_data_0;
+
+static int ov5640_init_0(const struct device *dev)
+{
+	const struct ov5640_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	if (!gpio_is_ready_dt(&cfg->reset_gpio)) {
+		LOG_ERR("%s: device %s is not ready", dev->name, cfg->reset_gpio.port->name);
+		return -ENODEV;
+	}
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, powerdown_gpios)
+	if (!gpio_is_ready_dt(&cfg->powerdown_gpio)) {
+		LOG_ERR("%s: device %s is not ready", dev->name, cfg->powerdown_gpio.port->name);
+		return -ENODEV;
+	}
+#endif
+
+	return ov5640_init(dev);
+}
+
+DEVICE_DT_INST_DEFINE(0, &ov5640_init_0, NULL, &ov5640_data_0, &ov5640_cfg_0, POST_KERNEL,
```

Comment:
Why not make this driver multi-instance? For SOC IP blocks (where only one instance exists on silicon) I can see the point of keeping a driver single instance, but a user can easily purchase and potentially connect two cameras for their application, right?

Reply:
Hi @loicpoulain, do you know if there is a specific reason that the mt9m114 and other camera drivers were explicitly made as unique instance ?

Reply:
Don't get the answer from @loicpoulain, turned the driver into multi-instance though